app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt

```diff
@@ -68,81 +61,79 @@ class IP2Country private constructor(private val context: Context) {
-        val directory = File(context.applicationInfo.dataDir)
-        val file = File(directory, fileName)
-        if (directory.list().contains(fileName)) { return file }
-        val inputStream = context.assets.open("csv/$fileName")
-        val outputStream = FileOutputStream(file)
-        val buffer = ByteArray(1024)
-        while (true) {
-            val count = inputStream.read(buffer)
-            if (count < 0) { break }
-            outputStream.write(buffer, 0, count)
-        }
-        inputStream.close()
-        outputStream.close()
-        return file
-    }
-
-    private fun cacheCountryForIP(ip: String): String? {
-
+    internal fun cacheCountryForIP(ip: String): String? {
         // return early if cached
         countryNamesCache[ip]?.let { return it }
 
-        val ipInt = Ipv4Int(ip)
-        val bestMatchCountry = ipv4ToCountry.floorEntry(ipInt)?.let { (_, code) ->
-            if (code != null) {
-                countryToNames[code]
-            } else {
-                null
-            }
-        }
+        val ipInt = ipv4Int(ip)
+        val index = ips.fuzzyBinarySearch(ipInt)
+        val code = index?.let { codes[it] }
+        val bestMatchCountry = countryToNames[code]
 
-        if (bestMatchCountry != null) {
-            countryNamesCache[ip] = bestMatchCountry
-            return bestMatchCountry
-        } else {
-            Log.d("Loki","Country name for $ip couldn't be found")
-        }
-        return null
+        if (bestMatchCountry != null) countryNamesCache[ip] = bestMatchCountry
+        else Log.d("Loki","Country name for $ip couldn't be found")
+
+        return bestMatchCountry
     }
 
     private fun populateCacheIfNeeded() {
         ThreadUtils.queue {
+            val start = System.currentTimeMillis()
             OnionRequestAPI.paths.iterator().forEach { path ->
                 path.iterator().forEach { snode ->
                     cacheCountryForIP(snode.ip) // Preload if needed
                 }
             }
+            Log.d("Loki","Cache populated in ${System.currentTimeMillis() - start}ms")
             Broadcaster(context).broadcast("onionRequestPathCountriesLoaded")
-            Log.d("Loki", "Finished preloading onion request path countries.")
         }
     }
     // endregion
 }
 
+@OptIn(ExperimentalUnsignedTypes::class)
+private fun UIntArray.fuzzyBinarySearch(target: UInt): Int? {
+    if (isEmpty()) return null
+
+    var low = 0
+    var high = size - 1
+
+    while (low <= high) {
+        val mid = (low + high) / 2
+        val midValue = this[mid]
+
+        when {
+            midValue == target -> return mid // Exact match found
+            midValue < target -> low = mid + 1 // Search in the right half
+            else -> high = mid - 1 // Search in the left half
+        }
+    }
+
+    // If no exact match, return the largest index with value <= target
+    return if (high >= 0) high else null
+}
```

Comment:
Can't we use the Kotlin's `UIntArray.binarySearch`?

Reply:
`binarySearch` is looking for an exact match or will return `-1`. This is searching for the "Price is Right" match. Best without going over... or the opposite. So we should get an answer for any given IP even if it wasn't explicitly provided in the list.

Reply:
`binarySearch` doesn't simply return -1 though, it returns "where the item should have been put in order to maintain the order", in other words, it will give you info about the closest matched IP. Refer to the doc of `binarySearch` here:

> the index of the element, if it is contained in the array within the specified range; otherwise, the inverted insertion point (-insertion point - 1). The insertion point is defined as the index at which the element should be inserted, so that the array (or the specified subrange of array) still remains sorted.

Reply:
oh, amazing, thanks @SessionHero01, done.